[FLINK-7040] [REST] Add basics for REST communication

zentol opened the pull request:
## What is the purpose of the change

This PR implements the fundamentals for generic Client/Server REST communication that will be used for the client/cluster communication, the WebRuntimeMonitor and queryable state.

#### Endpoints

`Endpoints` are the main runtime component that have to be started before any communication can happen. Their primary purpose is setting up the underlying netty stack.

The `RestClientEndpoint` is a fully-functional class that provides an asynchronous API for sending requests/receiving responses based around `CompleteFutures`. Requests are sent in a synchronous fashion; a new request is only sent out after a response was received for the previous request (for simplicity).

The `RestServerEndpoint` is an abstract class that is very similar to the `WebRuntimeMonitor`. Implementations have to implement a single method `abstract Collection<? extends AbstractRestHandler<?, ?>> initializeHandlers();` that returns a collection of handlers that should be registered.

#### Messages

To send a request the client accepts 3 arguments:
* a `MessageHeaders`
* a `RequestBody`
* a `ParameterMapper`

`RequestBodies` represent the http message body, and will be converted to JSON using jackson-databind.

`ParameterMappers` are used to assemble the final url, including query and path parameters

`MessageHeaders` are stateless objects that define a link between requests and responses as well as provide meta-data about the communication for this particular pair. Headers have generic type arguments for requests and responses and provide some level of type-safeness; if the client and server use the same headers class the request/response type, url, http status codes etc. are all well-defined.
Essentially, headers provide a tight coupling between handlers, clients, requests and responses (provided that implementations don't define arbitrary headers on the fly).

For example, to define the communication for submitting a job one would define a `JobSubmitRequestBody` that contains the serialized job graph, a `JobSubmitResponseBody` containing the URL to track the job status and a `JobSubmitHeaders` class that define the HTTP method (POST), the url (e.g `/submit`, the response http status code (ACCEPTED).

#### Handlers

Someone has to deal with requests and send out responses, which is where the `AbstractRestHandler`  comes into play. This is an abstract class that manages all the netty stuff, and, just like the `MessageHeaders`, has generic type arguments for a `RequestBody` and `ResponseBody`. Only a single method must be implemented, which accepts a request and returns a response: `abstract CompletableFuture<HandlerResponse<P>> handleRequest(@Nonnull HandlerRequest<R> request);`

A `HandlerRequest` contains a `RequestBody` and maps for path/query parameters.
A `HandlerResponse` contains either a `ResponseBody`, or an `ErrorResponse`.

## Brief change log

  - Add `MessageHeaders`, `Request-/ResponseBody` for modeling messages
  - Add `AbstractRestHandler`, `HandlerRequest/-Response` for message processing
  - Add `RestClient-/-ServerEndpoint` that setup netty

## Verifying this change

This change added tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

zentol commented:
@tillrohrmann @kl0u Would be great if you could take a look.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java:
```diff
@@ -0,0 +1,263 @@
+		try {
+			// write request
+			Channel channel = bootstrap.connect(configuredTargetAddress, configuredTargetPort).sync().channel();
+			channel.writeAndFlush(httpRequest);
```
Here we could retrieve the ClientHandler via `channel.pipeline().get(ClientHandler.class)` and then get the future of this handler back.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java:
```diff
@@ -0,0 +1,263 @@
+		return responseFuture;
+	}
+
+	@ChannelHandler.Sharable
```
Let's not make it sharable and instead every handler has its own future it completes.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java:
```diff
@@ -0,0 +1,263 @@
+			if (msg instanceof FullHttpResponse) {
+				completeFuture((FullHttpResponse) msg);
+			} else {
+				LOG.error("Implementation error: Received a response that wasn't a FullHttpResponse.");
```
Shouldn't this fail the associated future?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java:
```diff
@@ -0,0 +1,263 @@
+				LOG.debug("Received response {}.", rawResponse);
+			} catch (JsonMappingException | JsonParseException je) {
+				LOG.error("Response was not valid JSON.", je);
+				return;
```
completeExceptionally the associated future.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java:
```diff
@@ -0,0 +1,263 @@
+			}
+
+			final ExpectedResponse<J> expectedResponse = (ExpectedResponse<J>) this.expectedResponse;
+			this.expectedResponse = null;
```
This kind of shadowing is confusing and should be avoided.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java:
```diff
@@ -0,0 +1,263 @@
+					// lets see if it is an ErrorResponse instead
+					try {
+						ErrorResponseBody error = objectMapper.treeToValue(rawResponse, ErrorResponseBody.class);
+						expectedResponse.responseFuture.completeExceptionally(new RuntimeException(error.errors.toString()));
```
Let's not use `RuntimeExceptions`. I think it's better to introduce a `RestClientException` which extends the `FlinkException`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java:
```diff
@@ -0,0 +1,263 @@
+						// if this fails it is either the expected type or response type was wrong, most likely caused
+						// by a client/search MessageHeaders mismatch
+						LOG.error("Received response was neither of the expected type ({}) nor an error. Response={}", expectedResponse.responseClass, rawResponse, jme);
+						expectedResponse.responseFuture.completeExceptionally(new RuntimeException("Response was neither of the expected type(" + expectedResponse.responseClass + ") nor an error.", jme));
```
Same here with the exception.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java:
```diff
@@ -0,0 +1,263 @@
+				return;
+			}
+
+			try {
```
Not sure why we need the outer try-catch block?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpointConfiguration.java:
```diff
@@ -0,0 +1,105 @@
+	public RestClientEndpointConfiguration(String targetRestEndpointAddress, int targetRestEndpointPort, SSLEngine sslEngine) {
+		this.targetRestEndpointAddress = targetRestEndpointAddress;
+		this.targetRestEndpointPort = targetRestEndpointPort;
+		this.sslEngine = sslEngine;
```
`Preconditions.checkNotNull` missing

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestServerEndpoint.java:
```diff
@@ -0,0 +1,172 @@
+	private ServerBootstrap bootstrap;
+	private Channel serverChannel;
+
+	public RestServerEndpoint(RestServerEndpointConfiguration configuration) {
```
check that `configuration` is not null.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestServerEndpoint.java:
```diff
@@ -0,0 +1,172 @@
+				bootstrap.childGroup().shutdownGracefully();
+			}
+		}
+		log.info("Rest endpoint shutdown complete.");
```
I think in order to say that we have to collect the termination futures of the groups and then wait on their completion. This should also have a timeout.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestServerEndpointConfiguration.java:
```diff
@@ -0,0 +1,103 @@
+	private RestServerEndpointConfiguration(String restBindAddress, int targetRestEndpointPort, SSLEngine sslEngine) {
+		this.restBindAddress = restBindAddress;
+		this.restBindPort = targetRestEndpointPort;
+		this.sslEngine = sslEngine;
```
`Preconditions.checkNotNull` missing

zentol replied:
the bind address and sslEngine are allowed to be null, in which case we bind to `localhost` / disable SSL.

tillrohrmann replied:
Then please annotate them with `@Nullable`

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractRestHandler.java:
```diff
@@ -0,0 +1,201 @@
+
+		try {
+			if (!(httpRequest instanceof FullHttpRequest)) {
+				log.error("Implementation error: Received a request that wasn't a FullHttpResponse.");
```
Why are non `FullHttpRequest` not allowed?

tillrohrmann replied:
typo: FullHttpRequest in the error message.

zentol replied:
It's not that they are not allowed, but we use a message aggregator in the pipeline that always returns a `FullHttpRequest`.

tillrohrmann replied:
Alright. Please add a comment explaining this.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractRestHandler.java:
```diff
@@ -0,0 +1,201 @@
+			ByteBuf msgContent = ((FullHttpRequest) httpRequest).content();
+
+			R request;
+			if (msgContent.capacity() == 0 || messageHeaders.getHttpMethod() == HttpMethodWrapper.GET) {
```
GET methods can have a body but it's not required.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractRestHandler.java:
```diff
@@ -0,0 +1,201 @@
+	 */
+	protected abstract CompletableFuture<HandlerResponse<P>> handleRequest(@Nonnull HandlerRequest<R> request);
+
+	private static <M extends MessageHeaders<R, P, ?>, R extends RequestBody, P extends ResponseBody> void sendResponse(M messageHeaders, P response, ChannelHandlerContext ctx, HttpRequest httpRequest) throws IOException {
```
We don't really need the `messageHeaders` here except for the status code. Wouldn't it make sense to pass only this value to the method?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/HandlerRequest.java:
```diff
@@ -0,0 +1,73 @@
+		this.pathParameters = pathParameters;
+		if (!queryParameters.isEmpty()) {
+			this.queryParameters = new HashMap<>();
+			queryParameters.forEach((key, value) -> this.queryParameters.put(key, value.get(0)));
```
Not sure whether we should copy this from the old `WebRuntimeMonitor`. I think there is a reason why the query parameters can have multiple value entries.

zentol replied:
While the HTTP specification doesn't forbid duplicate query parameters it also doesn't define how they should be handled. It is perfectly valid to handle the first/last parameter, or both either as a list or separately, the decision of which is entirely up to us.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/HandlerRequest.java:
```diff
@@ -0,0 +1,73 @@
+
+	public HandlerRequest(R requestBody, Map<String, String> pathParameters, Map<String, List<String>> queryParameters) {
+		this.requestBody = requestBody;
+		this.pathParameters = pathParameters;
```
`Preconditions.checkNotNull` missing.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/HandlerResponse.java:
```diff
@@ -0,0 +1,121 @@
+ *
+ * @param <P> type of successful response
+ */
+public final class HandlerResponse<P extends ResponseBody> {
```
I'm wondering whether we shouldn't make the `HandlerResponse` an interface which has two implementations `SuccessfulHandlerResponse` and `ErrorHandlerResponse`. We could add static methods to the `HandlerResponse` to create the respective instantiations. That way we would not have partly unused fields in the object and have type safety wrt the handler response.

zentol replied:
I can change that, but I don't see how this is superior regarding type safety.

tillrohrmann replied:
See the checks when accessing the different fields. This can easily throw an exception if the user did not properly check what's the type of the response.

zentol replied:
But even if we split this interface a user that gets a `HandlerResponse` still has to do the same checks as to whether it was successful or not. Unless you suggest to have an empty `HandlerResponse` interface and force users to do an instanceof+cast every time.

tillrohrmann replied:
Yes it would require casting the response. I think this is more explicit and prevents the user from shooting himself in the foot.

kl0u replied:
I think it would also be nice to allow the user to be able to specify a callback in the response which will be called when the response has been fully sent in the network. The reason for this is to be able to update metrics like "how long did the transfer take?" or "what is the progress so far (for big messages)". This would require adding the callback in the `HandlerResponse` which implements the `ChannelFutureListener` and in the `AbstractRestHandler.sendResponse()` register that listener with the `lastContentFuture`, if I understand correctly.

zentol replied:
@kl0u You can already do that by appending operations to the `CompletableFuture` that the handler returns, i.e. something like
```
resultFuture = ...
resultFuture.handle(userCallback)
return resultFuture
```

zentol replied:
@tillrohrmann I'd rather revisit an earlier suggestion of yours where we use a special checked exception (that we either throw or fail he future with).

tillrohrmann replied:
This would also be a good solution imo.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/PipelineErrorHandler.java:
```diff
@@ -0,0 +1,50 @@
+	@Override
+	protected void channelRead0(ChannelHandlerContext ctx, Object message) {
+		// we can't deal with this message. No one in the pipeline handled it. Log it.
+		logger.debug("Unknown message received: {}", message);
```
shouldn't we send an error message back to the client?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/RouterHandler.java:
```diff
@@ -0,0 +1,53 @@
+		try {
+			AbstractRestHandler.sendErrorResponse(new ErrorResponseBody("Not found."), HttpResponseStatus.NOT_FOUND, ctx, request);
+		} catch (IOException e) {
+			LOG.error("Critical error. Failed to send error response.", e);
```
Isn't the only reason for the `IOException` that we could not transform the `ErrorResponseBody` into a String? We should still be able to send an error message to the client then.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractRestHandler.java:
```diff
@@ -0,0 +1,201 @@
+						}
+					}
+				} catch (Exception e) {
+					log.error("Critical error while sending a response.", e);
```
I think the only reason for an exception here is the failed transformation of the (error)response to a string. If this should happen, then we should still be able to send something to the client. We should let him also know in this case.

tillrohrmann replied:
Actually this should happen in the `sendErrorResponse` and `sendResponse` methods.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/ErrorResponseBody.java:
```diff
@@ -0,0 +1,47 @@
+	static final String FIELD_NAME_ERRORS = "errors";
+
+	@JsonProperty(FIELD_NAME_ERRORS)
+	public final List<String> errors;
```
Why is this a list if we only instantiate it with a single string?

zentol replied:
The list constructor isn't meant to be private. The idea is to allow us to provide multiple error messages, which can be useful if a handler does multiple actions in parallel and multiple fail.

tillrohrmann replied:
But it's not yet used and somewhat misleading if it was initialized with a singleton list before. I suggest to remove it until needed.

zentol replied:
it's part of the API, so we either add it now or break it later.

tillrohrmann replied:
Alright, then leave it like it is.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/ParameterMapper.java:
```diff
@@ -0,0 +1,83 @@
+		pathParameters.forEach((parameter, value) -> {
+			int start = sb.indexOf(":" + parameter.getKey());
+			sb.replace(start, start + parameter.getKey().length() + 1, value);
+		});
```
Shouldn't we check that we have replace all placeholders and if not throw an exception?

zentol replied:
we can't check that in this static method since we neither know the original parameters nor can search for leftover parameters (since `:` is a valid character for a url). The only place where we could do it is in the client, but it wouldn't be able to tell whether the replace parts or correct anyway, so there might not be much value in checking.

tillrohrmann replied:
But we could, for example, count how many placeholders there are in the request URL and check that all of them have been replaced, couldn't we? Or is it also a valid request URL if a placeholder has not been replaced with a concrete value?

zentol replied:
Technically it could still be a valid URL.

My conclusion is that this implementation for handling parameters is inherently flawed.
* The mapper gets a set of parameters in a map but can decide to completely ignore it, (which is how most implementation will most likely look like)
* There is no good means of verifying whether parameters were set or not
* It doesn't it handle optional parameters, which is common for query parameters
* it is limited to the client side, on the server side we're still dealing with String:String maps

I think I've come up with something that addresses the above short-comings, hang tight.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/MessageHeaders.java:
```diff
@@ -0,0 +1,91 @@
+ * @param <R> request type
+ * @param <P> response type
+ */
+public interface MessageHeaders<R extends RequestBody, P extends ResponseBody, M extends ParameterMapper> {
```
For what do we need `M`?

zentol replied:
The idea is to have the headers define which parameter mapper is used, to prevent some unrelated mapper to be passed in.

tillrohrmann replied:
Alright, maybe this could be documented as part of the `@param <M>` description.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/ParameterMapper.java:
```diff
@@ -0,0 +1,83 @@
+	 * @return map containing the parameters and their associated value
+	 */
+	public Map<Parameter, String> mapQueryParameters(Set<Parameter> queryParameters) {
+		return Collections.emptyMap();
```
Why not leaving these methods abstract?

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestEndpointITCase.java:
```diff
@@ -0,0 +1,173 @@
+/**
+ * IT cases for {@link RestClientEndpoint} and {@link RestServerEndpoint}.
+ */
+public class RestEndpointITCase {
```
Let's extend the `TestLogger`

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/RestEndpointITCase.java:
```diff
@@ -0,0 +1,173 @@
+			clientEndpoint.sendRequest(new TestHeaders(), new TestParameterMapper(), new TestRequest()).get();
+		} catch (Exception e) {
+			clientEndpoint.shutdown();
+			serverEndpoint.shutdown();
```
Shouldn't this be in the finally block?

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/HandlerResponseTest.java:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Tests for {@link HandlerResponse}.
+ */
+public class HandlerResponseTest {
```
`TestLogger`

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/HandlerResponseTest.java:
```diff
@@ -0,0 +1,73 @@
+
+		try {
+			response.getErrorCode();
+			Assert.fail();
```
fail message would be helpful.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java:
```diff
@@ -155,38 +129,66 @@ public void shutdown() {
-		} catch (InterruptedException e) {
-			return FutureUtils.completedExceptionally(e);
+		synchronized (lock) {
+			CompletableFuture<P> responseFuture = ClientHandler.addHandlerForResponse(bootstrap, sslEngine, messageHeaders.getResponseClass());
```
Can't we do it by having a fixed channel initializer which always creates a new `ClientHandler`? Then we can retrieve the actual `ClientHandler` by `channel.pipeline.get(ClientHandler.class)`. And then we can gain access to the future.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java:
```diff
@@ -155,38 +129,66 @@ public void shutdown() {
-			channel.closeFuture();
-		} catch (InterruptedException e) {
-			return FutureUtils.completedExceptionally(e);
+		synchronized (lock) {
```
If we have a fixed channel initializer, then we don't need this lock here.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java:
```diff
@@ -155,38 +129,66 @@ public void shutdown() {
+				// write request
+				Channel channel = bootstrap.connect(configuredTargetAddress, configuredTargetPort).sync().channel();
+				channel.writeAndFlush(httpRequest);
+				channel.closeFuture();
```
What's the purpose of this call which obtains the close future but does not do anything with it?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java:
```diff
@@ -155,38 +129,66 @@ public void shutdown() {
+
+			try {
+				// write request
+				Channel channel = bootstrap.connect(configuredTargetAddress, configuredTargetPort).sync().channel();
```
Can we do the `sync` in a future, then it's not blocking?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java:
```diff
@@ -155,38 +129,66 @@ public void shutdown() {
+		}
 
 		@Override
 		protected void channelRead0(ChannelHandlerContext ctx, Object msg) throws Exception {
```
Where do we actually close the connection?

tillrohrmann replied:
Then we should find it out.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClientEndpoint.java:
```diff
@@ -0,0 +1,263 @@
+			.channel(NioSocketChannel.class)
+			.handler(initializer);
+
+		LOG.info("Rest client endpoint started.");
```
I'd rather keep it this way to have the client be structured similarly as the server.

tillrohrmann replied:
I actually think it is a better design to do it in the constructor. That way we don't have the additional call to `start` which is necessary in the `RestServerEndpoint` case because of the `initializeHandlers` methods.

zentol replied:
its not strictly _necessary_ for the server since the `RestServerEndpoint` constructor could accept the list of handlers.

tillrohrmann replied:
True but wrt the current design it is.

zentol commented:
@tillrohrmann I've addressed most of your comments.

I have implemented the `HandlerResponse` as you suggested (casting option) for now, and include all query parameter as a list.

## Unaddressed comments:
* Cleaner shutdown of netty components in server
* Closing of netty connection (_somewhere_)
* RestEndpointITCase: client/server shutdown in finally block

##

## Parameter Rework
* Remove the `ParameterMapper` and related code
* `MessageHeaders` now has a generic `MessageParameters` argument
* `MessageParameters` is an abstract class that contains a collection of `MessageParameter`s that are suppported for the given request, and provides a utiltiy method for checking whether these parameters were resolved
* `MessageParameter` is also an abstract class, that defines the key, accepts a value (i.e. resolve the parameter), defines whether the parameter is mandatory/optional and whether it is a query/path parameter

### Usage

Let's say we want to have a request for getting the details of a job, with a path parameter for the job id :

We first define a path parameter for the job id (or reuse an existing one!):

```
public final class JobIDPathParameter extends MessagePathParameter {
	public JobIDPathParameter () {
		super("jobid", MessageParameterRequisiteness.MANDATORY);
	}

	public void resolve(JobID jobID) {
		super.resolve(jobID.toString());
	}
}
```

We then define the sum of all parameters for the request:

```
public final class JobDetailsParameters extends MessageParameters {
	private final JobIDPathParameter jobID= new JobIDPathParameter ();

	@Override
	public Collection<MessageParameter> getParameters() {
		return Collections.singleton(jobID);
	}
}
```

And finally, we include it in the JobDetailsHeaders:
```
public final class JobDetailsHeaders extends MessageHeaders<JobDetailsRequest, JobDetailsResponse, JobDetailsMessageParameters> {
...
```

The usage for the client would then look like this:
```
JobID jobID = ...
JobDetailsParameters parameters = headers.getUnresolvedParameters();
JobDetailsParameters.jobID.resolve(jobID);
client.sendRequest(headers, parameters, request);
```

zentol commented:
@tillrohrmann I've reworked the client response parsing: the `ClientHandler` now only parses the response as JSON, whereas the conversion to the respective `ResponseBody` is handled in a future declared in `RestClientHandler#submitRequest()`. With this the `ClientHandler` is once again immutable and we don't have to do any casts.

I've also reworked how handlers communicate failures: The `HandlerResponse` classes have been removed, and handlers may now throw a (checked) `RestHandlerException` or complete the future with an exception (with special treating for RestHandlerExceptions). Unchecked or non-RestHandlerExceptions are still treated as implementation errors and will return a "500 Internal Server Error".

zentol commented:
@tillrohrmann I've address the remaining comments.

tillrohrmann commented:
One last change I'd like to apply before merging is to let the user provide the target address and target port to the method `sendRequest` instead of to the `RestClientEndpoint` constructor. That way, we can reuse the client to talk to multiple REST endpoints. I'll do the changes.

kl0u commented:
Thanks a lot @tillrohrmann ! This is a pretty useful API addition.
